File: src/main/java/com/yumii/core/Money.java
package com.yumii.core;

public class Money implements Transactional {
  private double _amount;

  public Money(double amount) {
    _amount = amount;
  }

  public Transaction deposit(double amount) {
    _amount += amount;
    return Transaction.deposit(amount);
  }

  public Transaction withdraw(double amount) throws IllegalArgumentException {
    if (amount > _amount) {
      throw new IllegalArgumentException("Insufficient balance.");
    }
    _amount -= amount;
    return Transaction.withdraw(amount);
  }

  public static Money from(double amount) {
    return new Money(amount);
  }

  public double getAmount() {
    return this._amount;
  }
}

File: src/core/money.rs
use std::io::{Error, ErrorKind::Other};

use super::transaction::{Transaction, Transactional};

pub struct Money {
    amount: u16,
}

impl Money {
    pub fn from(amount: u16) -> Money {
        Self { amount }
    }

    pub fn amount(&self) -> u16 {
        self.amount
    }
}

impl Transactional for Money {
    fn deposit(&mut self, amount: u16) -> Transaction {
        self.amount += amount;
        Transaction::deposit(amount)
    }

    fn withdraw(&mut self, amount: u16) -> Result<Transaction, Error> {
        if self.amount < amount {
            return Err(Error::new(
                Other,
                "amount to take out must be lesser than actual",
            ));
        }
        self.amount -= amount;
        Ok(Transaction::withdraw(amount))
    }
}
